flink-runtime/src/main/java/org/apache/flink/runtime/source/coordinator/SourceCoordinatorContext.java

```diff
@@ -220,8 +223,20 @@ void sendEventToSourceOperator(int subtaskId, OperatorEvent event) {
                 String.format("Failed to send event %s to subtask %d", event, subtaskId));
     }
 
-    ScheduledExecutorService getCoordinatorExecutor() {
-        return coordinatorExecutor;
+    void sendEventToSourceOperatorIfTaskReady(int subtaskId, OperatorEvent event) {
+        checkSubtaskIndex(subtaskId);
+
+        callInCoordinatorThread(
+                () -> {
+                    final OperatorCoordinator.SubtaskGateway gateway =
+                            subtaskGateways.getOnlyGatewayAndNotCheckReady(subtaskId);
+                    if (gateway != null) {
+                        gateway.sendEvent(event);
+                    }
```

Comment:
I think there might be a race condition. Coordinator might run this check successfully and start sending an event, while simultaneously the receiver starts restarting? Shouldn't we have a `try/catch` here and actually checking if the exception is "retry-able", so things like "subtask has failed" or "subtask not ready" or "task is initialising"? 🤔

Reply:
There maybe no race condition, because all the operations that get/update SourceCoordinatorContext::subtaskGateways will execute in CoordinatorExecutor(just a single thread executor)~

Reply:
Hi @LoveHeat , as I understand, the operations of subtaskGateways isn't race condition. Piotr means when you intend to call `gateway.sendEvent(event)`, the gateway must not have a race condition. But the corresponding subtask may be restarting, causing gateway.sendEvent(event) to fail.

Please correct me if I'm wrong, thanks~

Reply:
> But the corresponding subtask may be restarting, causing gateway.sendEvent(event) to fail, right?

Yes, that's my concern.

Re your 2nd and 3rd paragraphs:
If `sendEvent` fails, it will throw an exception, that will bubble up until the catch-all safety net in `schedulePeriodTask`, so that will cause the job to failover. Which I guess is currently mostly fine, unless that source has been used in batch job with multiple failover regions. In that case expected behaviour should be that only subtasks belonging to the failover region with source subtasks should restart, not whole job. That's my main worry with this race condition. That's why I suggested:

> Shouldn't we have a try/catch here and actually checking if the exception is "retry-able", so things like "subtask has failed" or "subtask not ready" or "task is initialising"? 🤔

Reply:
but `gateway.sendEvent()` only throw three potential exception:
1. [task is not ready](https://github.com/apache/flink/blob/master/flink-runtime/src/main/java/org/apache/flink/runtime/operators/coordination/SubtaskGatewayImpl.java#L92), i think this exception will not thrown for ever, because we have already check task ready before call sendEvent()
2. [serialize event failed](https://github.com/apache/flink/blob/master/flink-runtime/src/main/java/org/apache/flink/runtime/operators/coordination/SubtaskGatewayImpl.java#L101), i think this exception  can not retry, should just fail job(but i don't known whether fail job for batch job is ok? )
3. just like: `if sendEvent fails, it won't throw exception directly and call the subtaskAccess.triggerTaskFailover.`

For other exception (like oom , jvm error)  is also can not be retried

(correct me if i misunderstand your opinion )

@pnowojski @1996fanrui

Reply:
Ahh, I think you are right. Thanks for pointing this out. In that case this code be fine